File: src/cmcandy/java_language_answers/_0042.java
package cmcandy.java_language_answers;

/**
 * Created by CHEN on 2020/5/14.
 */
public class _0042 {

    static class Solution {
        public int trap(int[] height) {

            int len = height.length;
            if (len < 3) return 0;
            int max = 0;
            int sum = 0;
            //动态规划，从左往右的最大值
            int[] left = new int[len];
            //从右往左的最大值
            int[] right = new int[len];

            for (int i = 0; i < len; i++) {
                if (height[i] > max) max = height[i];
                left[i] = max;
            }
            max = 0;
            for (int i = len - 1; i >= 0; i--) {
                if (height[i] > max) max = height[i];
                right[i] = max;
            }
            //两边取最小值计算和height的差值，即可找到该块的积水量
            for (int i = 0; i < len; i++) {
                int tmp = Math.min(left[i], right[i]);
                sum += tmp - height[i];
            }
            return sum;
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        System.out.println(solution.trap(new int[]{0,1,0,2,1,0,1,3,2,1,2,1} ) );
    }

}

File: src/cmcandy/C_Language_Answers/_0042.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace ::std;

class Solution
{
public:
    int trap(vector<int> &height)
    {
        //双指针
        int len = height.size();
        if (len < 3)
            return 0;
        //左边最高、右边最高
        int left = 0, right = len - 1;
        int li = 0, ri = right; //左右指针
        int sum = 0;
        int jud = 0; //0=li++ 1=ri--
        //如果右边比较埃，则右边先开始
        if (height[left] > height[right])
            jud = 1;
        while (li < ri)
        {
            //若当前列<min(l,r),加上差值
            //若当前列至少比其中一个大，则替换小的
            if (jud == 0)
            {
                if (height[++li] > height[left])
                {
                    left = li;
                    //如果是比右边高，则换个方向
                    if (height[li] > height[right])
                        jud = 1;
                }
                else
                {
                    sum += height[left] - height[li];
                }
            }
            else
            {
                if (height[--ri] > height[right])
                {
                    right = ri;
                    //如果是比左边高，则换个方向
                    if (height[ri] > height[left])
                        jud = 0;
                }
                else
                {
                    sum += height[right] - height[ri];
                }
            }
        }
        return sum;
    }
};